storage/badger/monotonic_consumer_progress.go

```diff
@@ -0,0 +1,54 @@
+package badger
+
+import (
+	"fmt"
+
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/onflow/flow-go/module/counters"
+)
+
+// MonotonicConsumerProgress represents the consumer progress with strict monotonic counter.
+type MonotonicConsumerProgress struct {
+	consumerProgress *ConsumerProgress
+
+	// used to skip heights that are lower than the current height
+	counter counters.StrictMonotonousCounter
+}
+
+// NewMonotonicConsumerProgress creates a new MonotonicConsumerProgress which inserts the default
+// processed index to the storage layer and creates new counter with defaultIndex value.
+//
+// No errors are expected during normal operation.
+func NewMonotonicConsumerProgress(db *badger.DB, consumer string, defaultIndex uint64) (*MonotonicConsumerProgress, error) {
+	m := &MonotonicConsumerProgress{
+		consumerProgress: NewConsumerProgress(db, consumer),
+		counter:          counters.NewMonotonousCounter(defaultIndex),
+	}
+
+	err := m.consumerProgress.InitProcessedIndex(defaultIndex)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
this should handle the 2 initialization cases where the db is empty or populated.

This is how it's done in the jobqueue:
https://github.com/onflow/flow-go/blob/9cc82c70399731de2dad6694d3bf68cc07fc38e6/module/jobqueue/consumer.go#L90-L105

the counter should then be populated using the existing value if it's set, or otherwise the default

Reply:
[8ad035c](https://github.com/onflow/flow-go/pull/5752/commits/8ad035cb2ee7386b451cc28fd1740487aef30217)